File: golang/rp_add_contact.go
//usr/bin/go run $0 $@; exit $?

// Program: rp_add_contact.go
// Usage: go run rp_add_contact.go
// Purpose: Go language sample client program for RackPing Monitoring API 2.0
// Version: 1.0
// Copyright: RackPing USA 2020
// Env: Go 1.13.14 or higher
// Returns: exit status is non-zero on failure
// Notes: first set envariables with: source ../set.sh

package main

import (
        "bytes"
        "encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
        "net/http/httputil"
	"os"
        "strconv"
	"time"
)

func debug(data []byte, err error) {
    if err == nil {
       fmt.Printf("%s\n\n", data)
    } else {
       log.Fatalf("%s\n\n", err)
    }
}

// global variables for communicating with callback redirect function redirectPolicyFunc()
var g_username string = ""
var g_password string = ""
var g_api_key  string = ""

func redirectPolicyFunc(req *http.Request, via []*http.Request) error {
   req.SetBasicAuth(g_username, g_password)
   req.Header.Set("app-key", g_api_key)
   return nil
}

func main() {
   g_username = os.Getenv("RP_USER")
   g_password = os.Getenv("RP_PASSWORD")
   g_api_key  = os.Getenv("RP_API_KEY")

   if g_username == "" {
      fmt.Printf("%s\n", "error: do 'source ../set.sh' first.")
      os.Exit(1)
   }

   var domain   string = os.Getenv("RP_DOMAIN")
   var base_url string = os.Getenv("RP_BASE_URL")
   var scheme   string = os.Getenv("RP_SCHEME")

   var Debug bool = (os.Getenv("RP_DEBUG") == "1")

   // see https://golang.org/pkg/time/ for how time casts work
   timeout, err := strconv.Atoi(os.Getenv("RP_TIMEOUT"))
   if err != nil {
      log.Fatal(err)
   }

   var url string = scheme + domain + base_url

   client := &http.Client{
      Timeout: time.Duration(timeout) * time.Second,
      CheckRedirect: redirectPolicyFunc,
   }

   {
      // if you want to understand this type definition, see the following links:
      // - https://gobyexample.com/json
      // - https://golang.org/pkg/encoding/json/#Marshal

      type RackPing struct {
           First       string `json:"first"`
           Last        string `json:"last"`
           Email       string `json:"email"`
           Role        string `json:"role"`
           Cellphone   string `json:"cellphone"`
           Countrycode int    `json:"countrycode"`
           Countryiso  string `json:"countryiso"`
      }

      m := RackPing{
         First:       "John",
         Last:        "Doe",
         Email:       "john.doe@example.com",
         Role:        "O",
         Cellphone:   "408 555 1212",
         Countrycode: 1,
         Countryiso:  "US" } // leave bracket on same line as last field

      b, err := json.Marshal(m)
      data := bytes.NewBuffer(b)

      println("Add one contact")

      req, err := http.NewRequest("POST", url + "/contacts", data)

      req.Header.Set("Content-Type", "application/json")
      req.SetBasicAuth(g_username, g_password)
      req.Header.Set("app-key", g_api_key)
      if Debug { debug(httputil.DumpRequestOut(req, true)) }

      res, err := client.Do(req)
      if err != nil {
         log.Fatal(err)
         debug(httputil.DumpResponse(res, true))
      } else {
         defer res.Body.Close()
         if Debug { debug(httputil.DumpResponse(res, true)) }
         body, err := ioutil.ReadAll(res.Body)
         if err == nil {
            fmt.Printf("%s\n", body)
         }
      }

      if Debug { debug(httputil.DumpResponse(res, true)) }
      println("HTTP status code is", res.StatusCode)
   }

   os.Exit(0)
}

File: node.js/rp_add_contact.js
#!/usr/bin/env node

// Program: rp_add_contact.js
// Usage: ./rp_add_contact.js
// Purpose: node.js language sample client program for RackPing Monitoring API 2.0
// Date: 2021 02 12
// Version: 1.0
// Copyright: RackPing USA 2020
// Env: Perl5
// Returns: exit status is non-zero on failure
// Note: First set the envariables with: source ../set.sh

const https = require("https");

   var api_key  = process.env.RP_API_KEY;
   if (api_key == "") {
      console.log("error: do source ../set.sh first");
      process.exit(1);
   }

   var host      = process.env.RP_DOMAIN;
   var base_url  = process.env.RP_BASE_URL;
   var user      = process.env.RP_USER;
   var password  = process.env.RP_PASSWORD;
   var timeout   = process.env.RP_TIMEOUT * 1000;
   var redirects = process.env.RP_REDIRECTS;
   var useragent = process.env.RP_USERAGENT;
   var debug     = process.env.RP_DEBUG;

// start of user settings

   var pw = '';
   for (i = 1; i <= 10; i++) {                   // sample code to assign a random password (in plaintext for password reminder email)
       pw += between(0, 15).toString(16);
   }

const data = JSON.stringify({
   'first':       'John',
   'last':        'Doe',
   'email':       'john.doe+' + api_key + '@example.com',
   'role':        'O',                           // user role: A = Admin, O = Operator, B = Billing
   'cellphone':   '408 555 1212',
   'countrycode': '1',                           // numeric telephone country code prefix
   'countryiso':  'US',                          // 2-letter country ISO code
   'lang':         'en',                         // language for new user and password reminder email
   'timezone':     'America/New_York',
   'sendemail':    '0',                          // send password reminder email [0|1]
   'alertable':    'N',                          // enable paid AlertPro user [Y|N]
   'pw':           pw                            // user password. blank disables password reminder email.
});

// end of user settings

   const options = {
       "method": "POST",
       "hostname": host,
       "port": 443,
       "path": base_url + "/contacts",
       "timeout": timeout,
//     "redirects": redirects,
       "user-agent": useragent,
       "headers": {
           "Accept":         "application/json",
           "Accept-Charset": "utf-8",
           "Content-type":   "application/json",
           "app-key":        api_key,
           "cache-control":  "no-cache",
           'Authorization':  'Basic ' + new Buffer(user + ':' + password).toString('base64')
       }
   }

   console.error("info: add one contact:");

   const req = https.request(options, function(res) {
       var chunks = [];

       res.on("data", function (chunk) {
           chunks.push(chunk);
       });

       res.on("end", function() {
           if (res.statusCode != 200) {
              console.log("error: status code", res.statusCode);
              process.exit(1);
           }

           var body = Buffer.concat(chunks);
           console.log(body.toString());
       });
   });

   req.write(data)
   req.end();

function between(min, max) {  
  return Math.floor(
    Math.random() * (max - min) + min
  )
}
